armory/instrument/instrument.py

```diff
@@ -0,0 +1,670 @@
+        stage_filter = filt[:-1].strip()
+        # tokens = [x.strip() for x in filt.split(",")]
+    else:
+        stage_filter = None
+    probe_variable = arg
+
+    return probe_variable, stage_filter
+
+
+class ProbeMapper:
+    """
+    Map from probe outputs to meters
+    """
+
+    def __init__(self):
+        # nested dicts - {probe_variable: {stage_filter: [(meter, arg)]}}
+        self.probe_filter_meter_arg = {}
+
+    def __len__(self):
+        """
+        Return the number of (meter, arg) pairs
+        """
+        count = 0
+        for probe_variable, filter_map in self.probe_filter_meter_arg.items():
+            for stage_filter, meters_args in filter_map.items():
+                count += len(meters_args)
+        return count
+
+    def __str__(self):
+        return f"{type(self)} : {self.probe_filter_meter_arg}"
+
+    def connect_meter(self, meter):
+        """
+        Connect meter to probes; idempotent
+        """
+        for arg in meter.get_arg_names():
+            probe_variable, stage_filter = process_meter_arg(arg)
+            if probe_variable not in self.probe_filter_meter_arg:
+                self.probe_filter_meter_arg[probe_variable] = {}
+            filter_map = self.probe_filter_meter_arg[probe_variable]
+            if stage_filter not in filter_map:
+                filter_map[stage_filter] = []
+            meters_args = filter_map[stage_filter]
+            if (meter, arg) in meters_args:
+                log.warning(
+                    f"(meter, arg) pair ({meter}, {arg}) already connected, not adding"
+                )
+            else:
+                meters_args.append((meter, arg))
+
+    def disconnect_meter(self, meter):
+        """
+        Disconnect meter from probes; idempotent
+        """
+        for arg in meter.get_arg_names():
+            probe_variable, stage_filter = process_meter_arg(arg)
+            if probe_variable not in self.probe_filter_meter_arg:
+                continue
+            filter_map = self.probe_filter_meter_arg[probe_variable]
+
+            if stage_filter not in filter_map:
+                continue
+            meters_args = filter_map[stage_filter]
+
+            if (meter, arg) in meters_args:
+                meters_args.remove((meter, arg))
+                if not meters_args:
+                    filter_map.pop(stage_filter)
+                if not filter_map:
+                    self.probe_filter_meter_arg.pop(probe_variable)
+
+    def map_to_meters_args(self, probe_variable, stage):
+        """
+        Return a list of (meter, arg) that are using the current probe_variable
+        """
+        filter_map = self.probe_filter_meter_arg.get(probe_variable, {})
+        meters = filter_map.get(stage, [])
+        meters.extend(filter_map.get(None, []))  # no stage filter (default)
+        return meters
+
+
+class Context:  # NOTE: may need to rename to something like Experiment or Procedure
```

Comment:
I think I agree that `Context` could use a rename, but am struggling to think of what would be best. `MetricRecorder`? Something that better captures the holistic nature of the class

Reply:
More discussion later.